deeppavlov/models/elmo/elmo_model.py

```diff
@@ -0,0 +1,732 @@
+            # Have added a special 0 index for padding not present
+            # in the original model.
+            embed_weights = fin[varname_in_file][...]
+            weights = np.zeros(
+                (embed_weights.shape[0] + 1, embed_weights.shape[1]),
+                dtype=DTYPE
+            )
+            weights[1:, :] = embed_weights
+    else:
+        with h5py.File(weight_file, 'r') as fin:
+            if varname_in_file == 'char_embed':
+                # Have added a special 0 index for padding not present
+                # in the original model.
+                char_embed_weights = fin[varname_in_file][...]
+                weights = np.zeros(
+                    (char_embed_weights.shape[0] + 1,
+                     char_embed_weights.shape[1]),
+                    dtype=DTYPE
+                )
+                weights[1:, :] = char_embed_weights
+            else:
+                weights = fin[varname_in_file][...]
+
+    # Tensorflow initializers are callables that accept a shape parameter
+    # and some optional kwargs
+    def ret(shape, **kwargs):
+        if list(shape) != list(weights.shape):
+            raise ValueError(
+                "Invalid shape initializing {0}, got {1}, expected {2}".format(
+                    varname_in_file, shape, weights.shape)
+            )
+        return weights
+
+    return ret
+
+
+class BidirectionalLanguageModelGraph(object):
+    '''
+    Creates the computational graph and holds the ops necessary for runnint
+    a bidirectional language model
+    '''
+    def __init__(self, options, weight_file, ids_placeholder,
+                 use_character_inputs=True, embedding_weight_file=None,
+                 max_batch_size=128):
+
+        self.options = options
+        self._max_batch_size = max_batch_size
+        self.ids_placeholder = ids_placeholder
+        self.use_character_inputs = use_character_inputs
+
+        # this custom_getter will make all variables not trainable and
+        # override the default initializer
+        def custom_getter(getter, name, *args, **kwargs):
+            kwargs['trainable'] = False
+            kwargs['initializer'] = _pretrained_initializer(
+                name, weight_file, embedding_weight_file
+            )
+            return getter(name, *args, **kwargs)
+
+        if embedding_weight_file is not None:
+            # get the vocab size
+            with h5py.File(embedding_weight_file, 'r') as fin:
+                # +1 for padding
+                self._n_tokens_vocab = fin['embedding'].shape[0] + 1
+        else:
+            self._n_tokens_vocab = None
+
+        with tf.variable_scope('bilm', custom_getter=custom_getter):
+            self._build()
+
+    def _build(self):
+        if self.use_character_inputs:
+            self._build_word_char_embeddings()
+        else:
+            self._build_word_embeddings()
+        self._build_lstms()
+
+    def _build_word_char_embeddings(self):
```

Comment:
This function duplicates in `bilm_model.py`.

Reply:
No, to be precise. Checks, layers and etc. There are many small differences between functions.